lib/staggered_two_link_quda.cu

```diff
@@ -0,0 +1,67 @@
+#include <utility>
+#include <quda_internal.h>
+#include <gauge_field.h>
+#include <ks_improved_force.h>
+#include <tune_quda.h>
+#include <instantiate.h>
+#include <tunable_nd.h>
+#include <kernels/staggered_two_link.cuh>
+
+namespace quda
+{
+
+  using namespace staggered_quark_smearing;
+
+  template <typename Arg> class TwoLink_ : public TunableKernel3D
+  {
+    Arg &arg;
+    const GaugeField &outA;
+    unsigned int minThreads() const { return arg.threads.x; }
+
+  public:
+    TwoLink_(Arg &arg, const GaugeField &link, const GaugeField &outA) :
+      TunableKernel3D(link, 2, 4), arg(arg), outA(outA)
+    {
+      strcat(aux, (std::string(comm_dim_partitioned_string()) + "threads=" + std::to_string(arg.threads.x)).c_str());
+      strcat(aux, ",TWO_LINK");
+
+      apply(device::get_default_stream());
+    }
+
+    void apply(const qudaStream_t &stream)
+    {
+      TuneParam tp = tuneLaunch(*this, getTuning(), getVerbosity());
+      launch<TwoLink>(tp, stream, arg);
+    }
+
+    void preTune() { outA.backup(); }
+
+    void postTune() { outA.restore(); }
+
+    long long flops() const { return 2 * 4 * arg.threads.x * 198ll; }
+
+    long long bytes() const { return 2 * 4 * arg.threads.x * (2 * arg.link.Bytes() + arg.outA.Bytes()); }
+  };
+
+  template <typename real, int nColor, QudaReconstructType recon> struct ComputeTwoLink {
+    ComputeTwoLink(const GaugeField &link, GaugeField &newTwoLink)
+    {
+      TwoLinkArg<real, nColor, recon> arg(newTwoLink, link);
+      TwoLink_<decltype(arg)> twolnk(arg, link, newTwoLink);
+    }
+  };
+#if defined(GPU_STAGGERED_DIRAC) && defined(GPU_TWOLINK_GSMEAR)
+  void computeTwoLink(GaugeField &newTwoLink, const GaugeField &link)
+  {
+    checkNative(newTwoLink, link);
+    checkLocation(newTwoLink, link);
+    checkPrecision(newTwoLink, link);
+    instantiate<ComputeTwoLink, WilsonReconstruct>(link, newTwoLink);
```

Comment:
Should this be `WilsonReconstruct` here or is `StaggeredReconstruct` more appropriate?

Reply:
I dug into this and realized `WilsonReconstruct` does make sense at face value, since the smearing does indeed seem to be done with the non-phased links. @hwancheolJeong I did see you have a comment in your test executable and your MILC code that `recon-12` isn't working, though. If that's the case, can you replace the `WilsonReconstruct` with `ReconstructNo` (so it'll be recon-18 only) and file an issue for it in this repo so it's revisited in the future?

Reply:
thanks @weinbe2 I agree it makes sense to restrict the reconstruction here as well , (with a fixme note)